pkg/allocator/cache.go

```diff
@@ -274,3 +290,48 @@ func (c *cache) numEntries() int {
 	defer c.mutex.RUnlock()
 	return len(c.nextCache)
 }
+
+type AllocatorChangeKind string
+
+const (
+	AllocatorChangeSync   AllocatorChangeKind = "sync"
+	AllocatorChangeUpsert AllocatorChangeKind = "upsert"
+	AllocatorChangeDelete AllocatorChangeKind = "delete"
+)
+
+type AllocatorChange struct {
+	Kind AllocatorChangeKind
+	ID   idpool.ID
+	Key  AllocatorKey
+}
+
+// Observe the allocator changes. Conforms to stream.Observable.
+// Replays the current state of the cache when subscribing.
+func (c *cache) Observe(ctx context.Context, next func(AllocatorChange), complete func(error)) {
+	go func() {
+		// Wait until initial listing has completed before
+		// replaying the state.
+		select {
+		case <-c.listDone:
+		case <-ctx.Done():
+			complete(ctx.Err())
+			return
+		}
+
+		c.mutex.RLock()
+		defer c.mutex.RUnlock()
```

Comment:
When exactly does this get released and how does that relate to the overall stream lifecycle?

Reply:
It will get released as soon as the observation starts -> `c.changeSrc.Observe` won't block because it will start a new goroutine in any case - even if it's just to call `complete`.

Reply:
I see, so `c.changeSrc.Observe()` is actually the anonymous `Observe()` function defined in `stream.Multicast[...](...)` in its FuncObservable (via `c.ChangeSrc` function populated in the cache constructor with a Multicast stream creation call), which I agree will spawn its own goroutine to handle subsequent events and unblock this caller + release the readlock.